src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -551,6 +551,97 @@ public SolutionCompilationState WithProjectOutputFilePath(ProjectId projectId, s
             forkTracker: true);
     }
 
+    public SolutionCompilationState WithProjectAttributes(ProjectInfo.ProjectAttributes attributes)
+    {
+        var projectId = attributes.Id;
+        var oldProject = SolutionState.GetRequiredProjectState(projectId);
+
+        if (oldProject.ProjectInfo.Attributes.Language != attributes.Language)
+        {
+            throw new NotSupportedException(WorkspacesResources.Changing_project_language_is_not_supported);
+        }
+
+        if (oldProject.ProjectInfo.Attributes.IsSubmission != attributes.IsSubmission)
+        {
+            throw new NotSupportedException(WorkspacesResources.Changing_project_from_ordinary_to_interactive_submission_or_vice_versa_is_not_supported);
+        }
+
+        return
+             WithProjectName(projectId, attributes.Name)
+            .WithProjectAssemblyName(projectId, attributes.AssemblyName)
+            .WithProjectFilePath(projectId, attributes.FilePath)
+            .WithProjectOutputFilePath(projectId, attributes.OutputFilePath)
+            .WithProjectOutputRefFilePath(projectId, attributes.OutputRefFilePath)
+            .WithProjectCompilationOutputInfo(projectId, attributes.CompilationOutputInfo)
+            .WithProjectDefaultNamespace(projectId, attributes.DefaultNamespace)
+            .WithHasAllInformation(projectId, attributes.HasAllInformation)
+            .WithRunAnalyzers(projectId, attributes.RunAnalyzers)
+            .WithProjectChecksumAlgorithm(projectId, attributes.ChecksumAlgorithm);
+    }
+
+    public SolutionCompilationState WithProjectInfo(ProjectInfo info)
+    {
+        var projectId = info.Id;
+        var newState = WithProjectAttributes(info.Attributes)
+            .WithProjectCompilationOptions(projectId, info.CompilationOptions)
+            .WithProjectParseOptions(projectId, info.ParseOptions)
+            .WithProjectReferences(projectId, info.ProjectReferences)
+            .WithProjectMetadataReferences(projectId, info.MetadataReferences)
+            .WithProjectAnalyzerReferences(projectId, info.AnalyzerReferences);
+
+        var oldProjectState = SolutionState.GetRequiredProjectState(projectId);
+
+        using var _1 = ArrayBuilder<DocumentInfo>.GetInstance(out var addedDocumentInfos);
+        using var _2 = ArrayBuilder<DocumentId>.GetInstance(out var removedDocumentInfos);
```

Comment:
doc that thse are temp buffers that will get filled/cleared across each call to UpdateDOcuments.